### DISCUSSION TITLE
First-class `System.Action<...>` and `System.Func<...>`

### DISCUSSION THREAD

**Discussion by RenderMichael** [dotnet/csharplang#8445] [Language Ideas]
Currently, the following does not compile, citing an ambiguous method resolution issue:

```csharp
M(() => true);

static void M<T>(Func<T> d)
{
}

static void M<T>(MyDelegate<T> d)
{
}

delegate TActual MyDelegate<TActual>();
```

We have some custom delegates in our libraries, and we would like to support the "common" delegate types (as part of migrating away from custom delegates entirely). However, we cannot introduce overloads without causing issues, and the only way we can migrate is with a breaking change.

The `Action` and `Func` families are currently natural types for lambdas. It would be consistent and beneficial if they were considered "better" when there's multiple delegate types that match.

**Comment by BlinD-HuNTeR** [comment]
I suppose you can indeed introduce the new overloads and use the new "Overload Resolution Priority" feature of C# 13, to make those overloads be preferred.

**Reply by jnm2** [reply]
Yes! This is the approach Shouldly is taking in <https://github.com/shouldly/shouldly/pull/984>. Because <https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/attributes/general#overloadresolutionpriority-attribute> requires the consuming project to be using langversion 13, the new overloads will only be available when you're targeting .NET 9.

**Reply by RenderMichael** [reply]
It's true that this kind of feature could only be implemented in C# Next, where the overload revolution attribute already exists. This is more general goodness IMO.

**Reply by jnm2** [reply]
The real issue is actually not the existence of the attribute- like all compiler-recognized attributes, this can be easily fixed by declaring the attribute per project as internal. The real issue is that the compiler simply ignores the attribute unless langversion is at least 13.

To drive this point home, you can set langversion to 13 in a .NET Framework project and declare the OverloadResolutionPriorityAttribute class manually, and see the feature fully working.

**Reply by RenderMichael** [reply]
Sounds like user-exposed compiler attributes are a little scary.

I feel like that attribute is best used for obsoletion, and not to supplement better-ness rules. `Action` and `Func` are already the natural types of lambdas, so there's precedent to treating them with higher priority.

**Reply by jnm2** [reply]
Betterness was a goal for the design; ~~.NET 9 is using these attributes for betterness with string.Concat and collection expressions so that a higher-performing span overload can be added without causing overload ambiguities when people update their langversion.~~

I agree that there could be something there with Action/Func being special. But let's consider, that would likely only apply when users set their langversion to C# 14 (if we're very optimistic). Anyone not using .NET 10 would still run into overload resolution issues. That's similar to the end result of ORPA (OverloadResolutionPriorityAttribute), only ORPA has a less strict version constraint, and ORPA is already shipped, whereas implementing a preference for Action/Func may never rise to priority for engineering time over other C# priorities.

**Reply by jnm2** [reply]
I forgot, the statement about .NET 9 using ORPA for string.Concat isn't correct. <https://github.com/dotnet/csharplang/issues/8374> is what addresses that scenario. It is safe to say, though, that the scenarios we've been discussing are the scenarios that ORPA was designed for.

**Reply by 333fred** [reply]
> `Action` and `Func` are already the natural types of lambdas, so there's precedent to treating them with higher priority.

To my mind, this actually makes the non-standard delegate types more likely to be the right things to call. In either case, the language cannot make a decision. Only you, the API author, can decide that, and `OverloadResolutionPriorityAttribute` is intentionally designed for you to be able to make such a decision.

**Reply by jaredpar** [reply]
> Sounds like user-exposed compiler attributes are a little scary.

Not nearly as scary as changing overload resolution itself.
